lib/optimize/UglifyJsPlugin.js

```diff
@@ -117,6 +166,41 @@ class UglifyJsPlugin {
 						asset.__UglifyJsPlugin = compilation.assets[file] = (map ?
 							new SourceMapSource(stringifiedStream, file, JSON.parse(map), input, inputSourceMap) :
 							new RawSource(stringifiedStream));
+						if(extractedComments.length > 0) {
+							let commentsFile = options.extractComments.file || file + ".LICENSE";
+							if(typeof commentsFile === "function") {
+								commentsFile = commentsFile(file);
+							}
+
+							// Write extracted comments to commentsFile
+							const commentsSource = new RawSource(extractedComments.join("\n\n") + "\n");
+							if(commentsFile in compilation.assets) {
+								// commentsFile already exists, append new comments...
+								if(compilation.assets[commentsFile] instanceof ConcatSource) {
+									compilation.assets[commentsFile].add("\n");
+									compilation.assets[commentsFile].add(commentsSource);
+								} else {
+									compilation.assets[commentsFile] = new ConcatSource(
+										compilation.assets[commentsFile], "\n", commentsSource
+									);
+								}
+							} else {
+								compilation.assets[commentsFile] = commentsSource;
+							}
+
+							// Add a banner to the original file
+							if(options.extractComments.banner !== false) {
+								let banner = options.extractComments.banner || "For license information please see " + commentsFile;
+								if(typeof banner === "function") {
+									banner = banner(commentsFile);
+								}
+								if(banner) {
+									asset.__UglifyJsPlugin = compilation.assets[file] = new ConcatSource(
```

Comment:
only write to `asset.__UglifyJsPlugin = compilation.assets[file] = ` once in the UglifyJsPlugin. Use a local variable to keep the uglify result source and reassign this. Add `asset.__UglifyJsPlugin = compilation.assets[file] = source` to the end of the plugin.